File: design_patterns/C++/Behavioral/strategy.cxx
#include <iostream>
#include <vector>

class SortStrategy {
  public:
	virtual void sort(std::vector<int>& data) = 0;
};

class BubbleSortStrategy: public SortStrategy {
  public:
	void sort(std::vector<int>& data) {
		std::cout << "Sorting using bubble sort" << std::endl;
	}
};

class QuickSortStrategy: public SortStrategy {
  public:
	void sort(std::vector<int>& data) {
		std::cout << "Sorting using quick sort" << std::endl;
	}
};

class Sorter {
  protected:
	SortStrategy* sorter = nullptr;

  public:
	Sorter(SortStrategy* sorter) {
		this->sorter = sorter;
	}

	void sort(std::vector<int>& data) {
		return this->sorter->sort(data);
	}
};

int main() {
	std::vector<int> data1 = {1, 5, 4, 3, 2, 8};
	auto sorter1 = new Sorter(new BubbleSortStrategy());
	sorter1->sort(data1);
	std::vector<int> data2 = {1, 5, 4, 3, 2, 8};
	auto sorter2 = new Sorter(new QuickSortStrategy());
	sorter2->sort(data2);
	// =========================
	// Sorting using bubble sort
	// Sorting using quick sort
	// =========================
	return 0;
}

File: design_patterns/Python/Behavioral/strategy.py
class SortStrategy:

    def sort(self, data):
        raise NotImplementedError


class BubbleSortStrategy(SortStrategy):

    def sort(self, data):
        print('Sorting using bubble sort')
        return data


class QuickSortStrategy(SortStrategy):

    def sort(self, data):
        print('Sorting using quick sort')
        return data


class Sorter:

    def __init__(self, sorter):
        self.sorter = sorter

    def sort(self, data):
        return self.sorter.sort(data)


if __name__ == '__main__':

    data = [1, 5, 4, 3, 2, 8]
    sorter = Sorter(BubbleSortStrategy())
    sorter.sort(data)  # Sorting using bubble sort

    sorter = Sorter(QuickSortStrategy())
    sorter.sort(data)  # Sorting using quick sort
